File: lib_elevator_control/src/main/java/com/reeman/agv/elevator/mqtt/MqttClient.java
package com.reeman.agv.elevator.mqtt;

import com.hivemq.client.mqtt.datatypes.MqttQos;
import com.hivemq.client.mqtt.mqtt3.Mqtt3AsyncClient;
import com.hivemq.client.mqtt.mqtt3.Mqtt3BlockingClient;
import com.hivemq.client.mqtt.mqtt3.message.connect.Mqtt3Connect;
import com.hivemq.client.mqtt.mqtt3.message.publish.Mqtt3Publish;
import com.hivemq.client.mqtt.mqtt3.message.subscribe.suback.Mqtt3SubAck;
import com.hivemq.client.rx.FlowableWithSingle;
import com.reeman.agv.elevator.BuildConfig;
import com.reeman.agv.elevator.exception.CustomException;
import com.reeman.agv.elevator.state.Code;
import com.reeman.agv.elevator.state.Step;

import java.util.concurrent.TimeUnit;

import io.reactivex.rxjava3.core.Observable;
import timber.log.Timber;


public class MqttClient {

    private final String host;
    private final String clientId;
    private final String username;
    private final String password;
    private Mqtt3BlockingClient client;
    private final String topic;
    private int reconnectCount;
    private boolean isManualDisconnect;
    private CallBack callBack;
    private boolean connectSuccess = false;

    private boolean isAlreadyReceiveLeave;

    public void setAlreadyReceiveLeave(boolean alreadyReceiveLeave) {
        isAlreadyReceiveLeave = alreadyReceiveLeave;
    }

    public boolean isManualDisconnect() {
        return isManualDisconnect;
    }

    public MqttClient(String host, String clientId, String username, String password, String topic, CallBack callBack) {
        this.host = host;
        this.clientId = clientId;
        this.username = username;
        this.password = password;
        this.topic = topic;
        this.callBack = callBack;
    }

    public Observable<Integer> connect() {
        return Observable.create(emitter -> {
            isManualDisconnect = false;
            isAlreadyReceiveLeave = false;
            reconnectCount = 0;
            String[] split = host.replace("tcp://", "").split(":");
            if (split.length != 2)
                throw new CustomException(Step.CONNECT_MQTT, Code.PARSE_MQTT_HOST_FAILED);
            Timber.tag(BuildConfig.ELEVATOR_DIR).w(" host : " + split[0] + " , port : " + split[1]);
            Mqtt3Connect connect = Mqtt3Connect.builder()
                    .keepAlive(10)
                    .build();
            client = com.hivemq.client.mqtt.MqttClient.builder()
                    .useMqttVersion3()
                    .identifier(clientId)
                    .serverHost(split[0])
                    .serverPort(Integer.parseInt(split[1]))
                    .simpleAuth()
                    .username(username)
                    .password(password.getBytes())
                    .applySimpleAuth()
                    .addConnectedListener(context -> {
                        if (connectSuccess &&reconnectCount != 0) {
                            Timber.tag(BuildConfig.ELEVATOR_DIR).w("重新连接成功");
                            reconnectCount = 0;
                            return;
                        }
                        connectSuccess = true;
                        Timber.tag(BuildConfig.ELEVATOR_DIR).w("连接成功");
                        emitter.onNext(0);
                    })
                    .addDisconnectedListener(context -> {
                        Timber.tag(BuildConfig.ELEVATOR_DIR).w(context.getCause(),"连接断开");
                        if (isManualDisconnect || isAlreadyReceiveLeave || callBack == null) return;
                        if (reconnectCount < 3) {
                            Timber.tag(BuildConfig.ELEVATOR_DIR).w("断开连接,正在重连 : %s", reconnectCount);
                            context.getReconnector()
                                    .reconnect(true)
                                    .delay(++reconnectCount * 2L, TimeUnit.SECONDS);
                            return;
                        }
                        Timber.tag(BuildConfig.ELEVATOR_DIR).w("断开连接,超过最大重连次数");
                        callBack.onThrows(new CustomException(Step.CONNECT_MQTT, Code.CONNECT_MQTT_FAILED));
                    })
                    .buildBlocking();
            client.connect(connect);
        });


    }

    public void publish(String topic,String msg){
        client.publishWith().topic(topic).qos(MqttQos.AT_LEAST_ONCE).payload(msg.getBytes()).send();
    }


    public FlowableWithSingle<Mqtt3Publish, Mqtt3SubAck> subscribeToTopic() {
        return client.toRx().subscribePublishesWith()
                .topicFilter(topic)
                .qos(MqttQos.AT_LEAST_ONCE)
                .applySubscribe();
    }

    public void disconnect() {
        callBack = null;
        isManualDisconnect = true;
        if (client != null) {
            try {
                Mqtt3AsyncClient mqtt3AsyncClient = client.toAsync();
                mqtt3AsyncClient.unsubscribeWith().topicFilter(topic).send();
                mqtt3AsyncClient.disconnect();
            } catch (Exception e) {
                e.printStackTrace();
            }
            client = null;
        }
    }

    public interface CallBack {

        void onThrows(Throwable throwable);
    }

}

File: AGV_v1.9.0/lib_dispatch/src/main/java/com/reeman/dispatch/mqtt/MqttClient.kt
package com.reeman.dispatch.mqtt

import android.os.Build
import com.hivemq.client.mqtt.MqttClient
import com.hivemq.client.mqtt.datatypes.MqttQos
import com.hivemq.client.mqtt.mqtt3.Mqtt3AsyncClient
import com.hivemq.client.mqtt.mqtt3.exceptions.Mqtt3DisconnectException
import com.hivemq.client.mqtt.mqtt3.message.subscribe.Mqtt3Subscription
import com.hivemq.client.mqtt.mqtt5.exceptions.Mqtt5DisconnectException
import com.reeman.dispatch.BuildConfig
import com.reeman.dispatch.exception.InitializeException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import timber.log.Timber
import java.util.concurrent.TimeUnit

class MqttClient(
    private val host: String,
    private val port: Int,
    private val clientId: String,
    private val username: String,
    private val password: String,
    private val topicSub: String,
    private val retryDelaySeconds: Long = 2,
    private val maxRetries: Int = 30,
) {
    private val TAG = BuildConfig.DISPATCH_DIR

    private var mqttClient: Mqtt3AsyncClient? = null

    private var isReconnecting = false
    private var reconnectCount = 0

    fun isConnected() = mqttClient?.state?.isConnected == true

    fun connect(
        onConnectResult: (success: Boolean, throwable: Throwable?) -> Unit,
        onDisconnected: (retry: Boolean, reconnectCount: Int, throwable: Throwable) -> Unit,
        onReconnectSuccess: () -> Unit,
    ) {
        unsubscribe { _, _ -> }
        mqttClient = MqttClient.builder()
            .useMqttVersion3()
            .identifier(clientId)
            .serverHost(host)
            .serverPort(port)
            .simpleAuth()
            .username(username)
            .password(password.encodeToByteArray())
            .applySimpleAuth()
            .addConnectedListener {
                if (reconnectCount == 0) {
                    Timber.tag(TAG).w("连接成功")
                    return@addConnectedListener
                }
                Timber.tag(TAG).w("重连成功")
                reconnectCount = 0
                CoroutineScope(Dispatchers.Main).launch {
                    onReconnectSuccess()
                }

            }
            .addDisconnectedListener {
                Timber.tag(TAG).w(it.cause, "mqtt连接断开")
                if (it.cause is Mqtt3DisconnectException || it.cause is Mqtt5DisconnectException) return@addDisconnectedListener
                Timber.tag(TAG).w("$clientId 断开连接,重连次数: $reconnectCount, 最大重连次数: $maxRetries")
                val retry = reconnectCount <= maxRetries
                if (retry) {
                    Timber.tag(TAG).w("$retryDelaySeconds s后开始重连")
                    it.reconnector.reconnect(true)
                        .delay(retryDelaySeconds, TimeUnit.SECONDS)
                }
                isReconnecting = retry
                CoroutineScope(Dispatchers.Main).launch {
                    onDisconnected(retry, reconnectCount, it.cause)
                }
            }
            .buildAsync()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            mqttClient!!
                .connectWith()
                .keepAlive(5)
                .cleanSession(false)
                .send()
                .whenComplete { _, throwable: Throwable? ->
                    Timber.tag(TAG).d("$clientId 连接结果: ${isConnected()}")
                    CoroutineScope(Dispatchers.Main).launch {
                        onConnectResult(isConnected(), throwable)
                    }
                }
        }
    }

    fun publish(
        topic: String, payload: String,
        onPublishResult: (success: Boolean, throwable: Throwable?) -> Unit
    ) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            if (mqttClient == null) {
                onPublishResult(false, NullPointerException("MqttClient is null!"))
                return
            }
            mqttClient!!.publishWith()
                .topic(topic)
                .qos(MqttQos.AT_LEAST_ONCE)
                .payload(payload.toByteArray())
                .send()
                .whenComplete { _, throwable ->
                    if (throwable != null) {
                        onPublishResult(false, throwable)
                    } else {
                        onPublishResult(true, null)
                    }
                }
        }
    }

    fun subscribe(
        onSubscribeResult: (success: Boolean, throwable: Throwable?) -> Unit,
        onTopicReceived: (topic: String, payload: String) -> Unit,
    ) {
        if (mqttClient == null) {
            onSubscribeResult(false, InitializeException("mqtt client not initialized"))
            return
        }

        val subscription = Mqtt3Subscription.builder()
            .topicFilter(topicSub)
            .qos(MqttQos.AT_LEAST_ONCE)
            .build()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            mqttClient!!.subscribeWith()
                .addSubscription(subscription)
                .callback {
                    CoroutineScope(Dispatchers.Main).launch {
                        onTopicReceived(it.topic.toString(), String(it.payloadAsBytes))
                    }
                }
                .send()
                .whenComplete { _, throwable: Throwable? ->
                    Timber.tag(TAG).d("$clientId 订阅结果: ${throwable == null}")
                    CoroutineScope(Dispatchers.Main).launch {
                        onSubscribeResult(throwable == null, throwable)
                    }
                }
        }
    }

    private fun unsubscribe(
        onUnsubscribeResult: ((success: Boolean, throwable: Throwable?) -> Unit)? = null
    ) {
        if (mqttClient == null) {
            onUnsubscribeResult?.invoke(true,null)
            return
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            mqttClient!!.unsubscribeWith()
                .addTopicFilter(topicSub)
                .send()
                .whenComplete { _, throwable: Throwable? ->
                    Timber.tag(TAG).d("$clientId 取消订阅结果: ${throwable == null}")
                    onUnsubscribeResult?.invoke(throwable == null, throwable)
                }
        }

    }

    fun disconnect() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            if (isConnected()) {
                unsubscribe()
                mqttClient?.disconnect()
                mqttClient = null
                Timber.tag(TAG).d("$clientId 主动断开连接")
            }
        }
    }

}